yo/app/scripts/directives/copy-or-migrate-mappings.js

```diff
@@ -0,0 +1,176 @@
+            ],
+            enableColumnResize: false,
+            selectedItems: $scope.selectedProject,
+            multiSelect: false
+        };
+
+        function createMigrationStatusMessage(inputDataModelId, referenceProjectId, type) {
+            var actionMsg = type;
+            var itemMsg = 'mappings from reference project ' + referenceProjectId + ' to a new project';
+            var additionalItemMsg = '';
+
+            if(inputDataModelId) {
+
+                additionalItemMsg = 'with help of the schema from input data model ' + inputDataModelId;
+            }
+
+            return function(statusMsg) {
+                return [actionMsg, itemMsg, additionalItemMsg, statusMsg].join(' ');
+            };
+        }
+
+        $scope.copyMappings = function() {
+
+            var selectedProject = $scope.selectedProject[0];
+            var referenceProjectId = selectedProject.uuid;
+
+            var payload = {
+                reference_project: referenceProjectId
+            };
+
+            var finishMessage = createMigrationStatusMessage(null, referenceProjectId, 'Copying');
+
+            ngProgress.start();
+
+            var dataModelIdPromise = $q.defer();
+
+
+
+            if($scope.selectedDataModel.length > 0) {
+
+                dataModelIdPromise.resolve($scope.selectedDataModel.uuid);
+            } else {
+
+                // utilise input data model from project
+                ProjectResource.get({id: referenceProjectId}, function(project) {
+
+                    dataModelIdPromise.resolve(project.input_data_model.uuid);
+                }, dataModelIdPromise.reject);
+            }
+
+            dataModelIdPromise.promise.then(function(dataModelId) {
+
+                payload.input_data_model = dataModelId;
+
+                finishMessage = createMigrationStatusMessage(dataModelId, referenceProjectId, 'Copying');
+
+                return ProjectResource.createProjectWithHelpOfExistingEntities({}, payload).$promise;
+            }).then(function() {
+
+
+                ngProgress.complete();
+                showAlert.show($scope, 'info', finishMessage('successfully finished.'));
+
+                $scope.updateGridData();
+
+                isNewProjectsHasBeenCreated = true;
+            }, function(resp) {
+
+                ngProgress.complete();
+                showAlert.show($scope, 'danger', finishMessage('failed.'), 5000);
+
+                console.log(resp);
+
+                $window.alert(resp.message || resp.data.error);
+            });
+        };
+
+        $scope.migrateMappings = function() {
+
+
+        };
+
+        $scope.showHelp = function() {
+
+            $modal.open({
+                templateUrl: 'views/controllers/copy-or-migrate-mappings-help.html'
+            });
+        };
+
+        $scope.save = function() {
+
+            var result = {
+                newProjectsHasBeenCreated: isNewProjectsHasBeenCreated
```

Comment:
This is not very stable. One can click on `copy mappings` and on `ok` before the mapping was copied and this would return false, while the copy request is still underway.
A better way would be to make `isNewProjectsHasBeenCreated` a promise that resolved or rejected by `copyMappings` and that closes only when this promise is resolved. We can add some timeouting as well.